pkg/operator/ceph/cluster/mon/health.go

```diff
@@ -366,3 +312,127 @@ func removeMonitorFromQuorum(context *clusterd.Context, clusterName, name string
 	logger.Infof("removed monitor %s", name)
 	return nil
 }
+
+func (c *Cluster) resolveInvalidMonitorPlacement(desiredMonCount int) (bool, error) {
+	nodeChoice, err := c.findInvalidMonitorPlacement(desiredMonCount)
+	if err != nil {
+		return true, fmt.Errorf("failed to find invalid mon placement %+v", err)
+	}
+
+	// no violation found
+	if nodeChoice == nil {
+		return false, nil
+	}
+
+	for name, nodeInfo := range c.mapping.Node {
+		if nodeInfo.Name == nodeChoice.Node.Name {
+			logger.Infof("rebalance: rescheduling mon %s from node %s", name, nodeInfo.Name)
+			c.failMon(len(c.clusterInfo.Monitors), desiredMonCount, name)
+			return true, nil
+		} else {
+			logger.Warningf("rebalance: no mon pod found on node %s", nodeChoice.Node.Name)
+		}
+	}
+
+	return false, nil
+}
+
+func (c *Cluster) findInvalidMonitorPlacement(desiredMonCount int) (*NodeUsage, error) {
+	nodeZones, err := c.getNodeMonUsage()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node monitor usage. %+v", err)
+	}
+
+	// compute two helpful global flags:
+	//  - does an empty zone exist
+	//  - does an empty node exist
+	emptyZone := false
+	emptyNode := false
+	for zi := range nodeZones {
```

Comment:
Makes sense here, but elsewhere in this patch the index iteration is because the range-over-value construct creates value copies but we are returning pointers into the data structure. iterating over value references would be awesome